examples/object-oriented-cpp/dyaa.cpp

```diff
@@ -2,149 +2,176 @@
+    // generate a phase-space point
+    auto tmp = hadronic_pspgen(rng, 10.0, 7000.0);
+    auto s = tmp.s;
+    auto t = tmp.t;
+    auto u = tmp.u;
+    auto x1 = tmp.x1;
+    auto x2 = tmp.x2;
+    auto jacobian = tmp.jacobian;
+
+    double ptl = sqrt((t * u / s));
+    double mll = sqrt(s);
+    double yll = 0.5 * log(x1 / x2);
+    double ylp = fabs(yll + acosh(0.5 * mll / ptl));
+    double ylm = fabs(yll - acosh(0.5 * mll / ptl));
+
+    jacobian *= hbarc2 / calls;
+
+    // cuts for LO for the invariant-mass slice containing the
+    // Z-peak from CMSDY2D11
+    if ((ptl < 14.0) || (fabs(yll) > 2.4) || (ylp > 2.4) || (ylm > 2.4) ||
+        (mll < 60.0) || (mll > 120.0)) {
+      continue;
     }
 
-    // store some metadata in the grid
-    grid.set_key_value("events", "10000000");
+    auto weight = jacobian * int_photo(s, t, u);
+    double q2 = 90.0 * 90.0;
 
-    // read out the stored value and print it on stdout
-    const auto value = grid.get_key_value("events");
-    std::printf("Finished running %s events.\n", value.c_str());
-
-    // write the grid to disk - with `.lz4` suffix the grid is automatically LZ4 compressed
-    const std::string filename = "DY-LO-AA.pineappl.lz4";
-    grid.write(filename);
+    std::vector<double> ntuples = {q2, x1, x2};
+    grid.fill(0, fabs(yll), 0, ntuples, weight);
+  }
+}
 
-    std::printf("Generated %s containing a a -> l+ l-.\n\n"
-        "Try running (PDF sets must contain non-zero photon PDF):\n"
-        "  - pineappl convolve %s NNPDF31_nnlo_as_0118_luxqed\n"
-        "  - pineappl --silence-lhapdf plot %s NNPDF31_nnlo_as_0118_luxqed MSHT20qed_nnlo > plot_script.py\n"
-        "  - pineappl --help\n", filename.c_str(), filename.c_str(), filename.c_str());
+int main() {
+  // --- create a new `Channels` function for the $\gamma\gamma$ initial state
+  PineAPPL::Channels channels;
+  PineAPPL::SubChannelEntry subchannels;
+  subchannels.entry.push_back({{22, 22}, 1.0});
+  PineAPPL::ChannelsEntry channels_entry;
+  channels_entry.channels_entry.push_back(subchannels);
+  channels.add(channels_entry);
+
+  // --- Instatiate the Order object
+  // only LO $\alpha_\mathrm{s}^0 \alpha^2 \log^0(\xi_\mathrm{R})
+  // \log^0(\xi_\mathrm{F}) \log^0(\xi_\mathrm{A})$
+  std::vector<PineAPPL::OrderV1> orders = {PineAPPL::OrderV1{0, 2, 0, 0, 0}};
+
+  // --- Define the binning
+  // we bin in rapidity from 0 to 2.4 in steps of 0.1
+  std::vector<double> bins = {0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8,
+                              0.9, 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7,
+                              1.8, 1.9, 2.0, 2.1, 2.2, 2.3, 2.4};
+
+  // --- Construct the PineAPPL grid
+  PidBasis pid_basis = Evol;
+  std::vector<std::int32_t> pids = {2212, 2212};
+
+  // Define the types of convolutions
+  ConvType h1 = UnpolPDF;
+  ConvType h2 = UnpolPDF;
+  std::vector<ConvType> convolution_types = {h1, h2};
+
+  // Define the Kinematics
+  Kinematics scales = {Scale, 0};
```

Comment:
the type `Kinematics` is coming from the C-API, right? mmm ... I don't like that name floating around globally (because it might well be used [by others](https://github.com/felixhekhorn/LeProHQ/blob/5f0c8664db9b0c9cb3d483d62b68dd1ecba6526b/LeProHQ%2B%2B/include/FullyDiff/KinematicVars.hpp#L15)) ... can't we do something against? C has no namespaces so there is nothing we can do directly there, right? should we add the `#include <pineappl_capi>` _inside_ the `namespace PineAPPL`? a 2min google search says this is in general a bad idea - but I think for our case this is justified: because the two things are actually one and the same and there are bound together forever. If you break C you break also C++

Reply:
You can't use `namespace` in C because that's not a keyword in C. It will also probably break the entire thing, because if you rename symbols the linker won't find them anymore. However, I agree that this is a problem that must be solved for the all new enums. I suggest we use `[export.rename]` in `cbindgen.toml` to take of this:
```toml
"ConvType" = "pineappl_conv_type"
"InterpMeth" = "pineappl_interp_meth"
```
and so on and so forth, so basically converting CamelCase to snake_case with a `pineappl_` prefix (that's basically namespaces in C).

Reply:
I fully agree that the best way is to have a mapping in `cbindgen.toml`.

Reply:
Done in [8db10ab](https://github.com/NNPDF/pineappl/pull/328/commits/8db10abd10f3bd9a74aaddd79c2d7e81181f3ac7).